File: __tests__/mockjs_test.ml
open Jest
open ExpectJs

(* TODO: move to BS std lib *)
external bind : ('a -> 'b) -> 'c -> 'a -> ('a -> 'b) = "" [@@bs.send]
external bindThis : ('a -> 'b [@bs]) -> 'c -> ('a -> 'b [@bs]) = "bind" [@@bs.send]
external call : ('a -> 'b [@bs]) -> 'c -> 'a -> 'b = "" [@@bs.send]
let call self arg = call self () arg
external callThis : ('a -> 'b [@bs]) -> 'c -> 'a -> 'b = "call" [@@bs.send]
external call2 : ('a -> 'b -> 'c [@bs]) -> (_ [@bs.as 0]) -> 'a -> 'b -> 'c = "call" [@@bs.send]

let _ = 

describe "inferred_fn" (fun _ ->
  test "returns undefined" (fun _ ->
    let mockFn = JestJs.inferred_fn () in
    let fn = MockJs.fn mockFn in
    
    expect (call fn ()) |> toBeUndefined
  );
  
  test "black hole for argument type object" (fun _ ->
    let mockFn = JestJs.inferred_fn () in
    let fn = MockJs.fn mockFn in
    
    expect (call fn [%obj { property = 42 }]) |> toBeUndefined
  );
  
  test "black hole for argument type string" (fun _ ->
    let mockFn = JestJs.inferred_fn () in
    let fn = MockJs.fn mockFn in
    
    expect (call fn "some string") |> toBeUndefined
  );
  
  test "calls - records call arguments" (fun _ ->
    let mockFn = JestJs.inferred_fn () in
    let fn = MockJs.fn mockFn in
    
    let _ = call fn "first" in
    let _ = call fn "second" in
    let calls  = mockFn |> MockJs.calls in
    
    expect calls |> toEqual [| "first"; "second" |]
  );
  
  test "mockClear - resets calls" (fun _ ->
    let mockFn = JestJs.inferred_fn () in
    let fn = MockJs.fn mockFn in
    
    let before  = mockFn |> MockJs.calls in 
    let _ = (call fn 1, call fn 2) in
    let inbetween  = mockFn |> MockJs.calls in
    MockJs.mockClear mockFn;
    let after  = mockFn |> MockJs.calls in
    
    expect
      (before, inbetween, after)
    |> toEqual
      ([||], [| 1; 2 |], [||])
  );
  
  test "mockReset - resets calls" (fun _ ->
    let mockFn = JestJs.inferred_fn () in
    let fn = MockJs.fn mockFn in
    
    let before  = mockFn |> MockJs.calls in 
    let _ = (call fn 1, call fn 2) in
    let inbetween  = mockFn |> MockJs.calls in
    MockJs.mockReset mockFn;
    let after  = mockFn |> MockJs.calls in
    
    expect
      (before, inbetween, after)
    |> toEqual
      ([||], [| 1; 2 |], [||])
  );
  
  test "mockReset - resets implementations" (fun _ ->
    let mockFn = JestJs.inferred_fn () in
    mockFn |> MockJs.mockReturnValue (Js.Undefined.return 128) |> ignore;
    let fn = MockJs.fn mockFn in
    
    let before = call fn () in 
    MockJs.mockReset mockFn;
    let after = call fn () in 
    
    expect
      (before, after)
    |> toEqual
      (Js.Undefined.return 128, Js.Undefined.empty)
  );
  
  test "mockImplementation - sets implementation to use for subsequent invocations" (fun _ ->
    let mockFn = JestJs.inferred_fn () in
    let fn = MockJs.fn mockFn in
    
    let before = call fn 10 in
    mockFn |> MockJs.mockImplementation ((fun a -> Js.Undefined.return (string_of_int a)) [@bs]) |> ignore;
    
    expect
      (before, call fn 18, call fn 24)
    |> toEqual
      (Js.Undefined.empty, Js.Undefined.return "18", Js.Undefined.return "24")
  );
  
  test "mockImplementationOnce - queues implementation for one subsequent invocation" (fun _ ->
    let mockFn = JestJs.inferred_fn () in
    let fn = MockJs.fn mockFn in
    
    let before = call fn 10 in
    mockFn |> MockJs.mockImplementationOnce ((fun a -> Js.Undefined.return (string_of_int a)) [@bs]) |> ignore;
    mockFn |> MockJs.mockImplementationOnce ((fun a -> Js.Undefined.return (string_of_int (a * 2))) [@bs]) |> ignore;
    
    expect
      (before, call fn 18, call fn 24, call fn 12)
    |> toEqual
      (Js.Undefined.empty, Js.Undefined.return "18", Js.Undefined.return "48", Js.Undefined.empty)
  );
  
  test "mockReturnThis - returns `this` on subsequent invocations" (fun _ ->
    let mockFn = JestJs.inferred_fn () in
    let this = "this" in
    let fn = bindThis (mockFn |> MockJs.fn) this in
    
    let before = call fn () in
    mockFn |> MockJs.mockReturnThis |> ignore;
    
    expect
      (before, call fn (), call fn ())
    |> toEqual
      (Js.Undefined.empty, Js.Undefined.return this, Js.Undefined.return this)
  );
  
  test "mockReturnValue - returns given value on subsequent invocations" (fun _ ->
    let mockFn = JestJs.inferred_fn () in
    let fn = MockJs.fn mockFn in
    
    let before = call fn 10 in
    mockFn |> MockJs.mockReturnValue (Js.Undefined.return 146) |> ignore;
    
    expect
      (before, call fn 18, call fn 24)
    |> toEqual
      (Js.Undefined.empty, Js.Undefined.return 146, Js.Undefined.return 146)
  );
  
  test "mockReturnValueOnce - queues implementation for one subsequent invocation" (fun _ ->
    let mockFn = JestJs.inferred_fn () in
    let fn = MockJs.fn mockFn in
    
    let before = call fn 10 in
    mockFn |> MockJs.mockReturnValueOnce (Js.Undefined.return 29) |> ignore;
    mockFn |> MockJs.mockReturnValueOnce (Js.Undefined.return 41) |> ignore;
    
    expect
      (before, call fn 18, call fn 24, call fn 12)
    |> toEqual
      (Js.Undefined.empty, Js.Undefined.return 29, Js.Undefined.return 41, Js.Undefined.empty)
  );
  
  (*
  Skip.test "bindThis" (fun _ ->
    let fn = ((fun a -> string_of_int a) [@bs]) in
    let boundFn = bindThis fn "this" in
    
    expect (call boundFn () 2) |> toEqual "2"
  );
  *)
  
  (* TODO: Not applicable for function calls, should only be available for new calls
  test "mockClear - resets instances" (fun _ ->
    let mockFn = JestJs.inferred_fn () in
    
    let before  = mockFn |> MockJs.mock |> MockJs.instances in 
    let _ = (MockJs.fn mockFn (), MockJs.fn mockFn ()) in
    let inbetween  = mockFn |> MockJs.mock |> MockJs.instances in
    MockJs.mockClear mockFn; (* doesn't do anything? *)
    let after  = mockFn |> MockJs.mock |> MockJs.instances in
    
    expect
      (before, inbetween, after)
    |> toEqual
      ([||], [| Js.Undefined.empty; Js.Undefined.empty |], [||])
  );
  *)
);
  
describe "fn" (fun _ ->
  test "calls implementation" (fun _ ->
    let mockFn = JestJs.fn (fun a -> string_of_int a) in
    let fn = MockJs.fn mockFn in
    
    expect (fn 18) |> toBe "18"
  );
  
  test "calls - records call arguments" (fun _ ->
    let mockFn = JestJs.fn (fun a -> string_of_int a) in
    
    let _ = MockJs.fn mockFn 74 in
    let _ = MockJs.fn mockFn 89435 in
    let calls  = mockFn |> MockJs.calls in
    
    expect calls |> toEqual [| 74; 89435 |]
  );
  
  Skip.test "mockClear - resets calls" (fun _ ->
    let mockFn = JestJs.fn (fun a -> string_of_int a) in
    
    let before = mockFn |> MockJs.calls in 
    let _ = (MockJs.fn mockFn 1, MockJs.fn mockFn 2) in
    let inbetween = mockFn |> MockJs.calls in
    MockJs.mockClear mockFn;
    let after = mockFn |> MockJs.calls in
    
    expect
      (before, inbetween, after)
    |> toEqual
      ([||], [| 1; 2 |], [||])
  );
  
  test "mockReset - resets calls" (fun _ ->
    let mockFn = JestJs.fn (fun a -> string_of_int a) in
    let fn = MockJs.fn mockFn in
    
    let before  = mockFn |> MockJs.calls in 
    let _ = (fn 1, fn 2) in
    let inbetween  = mockFn |> MockJs.calls in
    MockJs.mockReset mockFn;
    let after  = mockFn |> MockJs.calls in
    
    expect
      (before, inbetween, after)
    |> toEqual
      ([||], [| 1; 2 |], [||])
  );
  
  (* TODO: Actually removes the original imlementation too, causing it to return undefined, which usually won't be a valid return value for the function type it mocks *)
  Skip.test "mockReset - resets implementations" (fun _ ->
    let mockFn = JestJs.fn (fun a -> string_of_int a) in
    mockFn |> MockJs.mockReturnValue "128" |> ignore;
    let fn = MockJs.fn mockFn in
    
    let before = fn 0 in 
    MockJs.mockReset mockFn;
    let after = fn 1 in 
    
    expect
      (before, after)
    |> toEqual
      ("128", "1")
  );
  
  test "mockImplementation - sets implementation to use for subsequent invocations" (fun _ ->
    let mockFn = JestJs.fn (fun a -> string_of_int a) in
    let fn = MockJs.fn mockFn in
    
    let before = fn 10 in
    mockFn |> MockJs.mockImplementation (fun a -> string_of_int (a * 2)) |> ignore;
    
    expect
      (before, fn 18, fn 24)
    |> toEqual
      ("10", "36", "48")
  );
  
  test "mockImplementationOnce - queues implementation for one subsequent invocation" (fun _ ->
    let mockFn = JestJs.fn (fun a -> string_of_int a) in
    let fn = MockJs.fn mockFn in
    
    let before = fn 10 in
    mockFn |> MockJs.mockImplementationOnce (fun a -> string_of_int (a * 3)) |> ignore;
    mockFn |> MockJs.mockImplementationOnce (fun a -> string_of_int (a * 2)) |> ignore;
    
    expect
      (before, fn 18, fn 24, fn 12)
    |> toEqual
      ("10", "54", "48", "12")
  );

  (* mockReturnThis doesn't make sense for native functions
  test "mockReturnThis - returns `this` on subsequent invocations" (fun _ ->
    let mockFn = JestJs.fn (fun a -> string_of_int a) in
    let this = "this" in
    let fn = bindThis (mockFn |> MockJs.fn) this in
    
    let before = fn () in
    mockFn |> MockJs.mockReturnThis |> ignore;
    
    expect
      (before, fn (), fn ())
    |> toEqual
      (Js.Undefined.empty, Js.Undefined.return this, Js.Undefined.return this)
  );
  *)
  
  test "mockReturnValue - returns given value on subsequent invocations" (fun _ ->
    let mockFn = JestJs.fn (fun a -> string_of_int a) in
    let fn = MockJs.fn mockFn in
    
    let before = fn 10 in
    mockFn |> MockJs.mockReturnValue "146" |> ignore;
    
    expect
      (before, fn 18, fn 24)
    |> toEqual
      ("10", "146", "146")
  );
  
  test "mockReturnValueOnce - queues implementation for one subsequent invocation" (fun _ ->
    let mockFn = JestJs.fn (fun a -> string_of_int a) in
    let fn = MockJs.fn mockFn in
    
    let before = fn 10 in
    mockFn |> MockJs.mockReturnValueOnce "29" |> ignore;
    mockFn |> MockJs.mockReturnValueOnce "41" |> ignore;
    
    expect
      (before, fn 18, fn 24, fn 12)
    |> toEqual
      ("10", "29", "41", "12")
  );
);

describe "fn2" (fun _ ->
  test "calls implementation" (fun _ ->
    let mockFn = JestJs.fn2 ((fun a b -> string_of_int (a + b)) [@bs]) in
    let fn = MockJs.fn mockFn in
    
    expect (call2 fn 18 24) |> toBe "42"
  );
);

(* TODO: depends on MockJs.make
describe "MockJs.make" (fun _ ->
  test "MockJs.make" (fun _ ->
    let mockFn = JestJs.fn (fun a -> string_of_int a) in
    
    let instance  = mockFn |> MockJs.make 4 in
    let instances  = mockFn |> MockJs.mock |> MockJs.instances in
    
    expect instances |> toEqual [| instance |]
  );
);
*)

File: lib/js/__tests__/mockjs_test.js
'use strict';

var Jest       = require("../src/jest.js");
var Curry      = require("bs-platform/lib/js/curry.js");
var Caml_int32 = require("bs-platform/lib/js/caml_int32.js");
var Pervasives = require("bs-platform/lib/js/pervasives.js");

function call(self, arg) {
  return self.call(/* () */0, arg);
}

describe("inferred_fn", (function () {
        Jest.test("returns undefined", (function () {
                var mockFn = jest.fn();
                return Jest.ExpectJs[/* toBeUndefined */29](Jest.ExpectJs[/* expect */0](mockFn.call(/* () */0, /* () */0)));
              }));
        Jest.test("black hole for argument type object", (function () {
                var mockFn = jest.fn();
                return Jest.ExpectJs[/* toBeUndefined */29](Jest.ExpectJs[/* expect */0](mockFn.call(/* () */0, {
                                    property: 42
                                  })));
              }));
        Jest.test("black hole for argument type string", (function () {
                var mockFn = jest.fn();
                return Jest.ExpectJs[/* toBeUndefined */29](Jest.ExpectJs[/* expect */0](mockFn.call(/* () */0, "some string")));
              }));
        Jest.test("calls - records call arguments", (function () {
                var mockFn = jest.fn();
                mockFn.call(/* () */0, "first");
                mockFn.call(/* () */0, "second");
                var calls = Jest.MockJs[/* calls */3](mockFn);
                return Jest.ExpectJs[/* toEqual */12](/* array */[
                            "first",
                            "second"
                          ], Jest.ExpectJs[/* expect */0](calls));
              }));
        Jest.test("instances - sanity check - is empty array", (function () {
                var mockFn = jest.fn();
                var instances = Jest.MockJs[/* instances */4](mockFn);
                return Jest.ExpectJs[/* toEqual */12](/* array */[], Jest.ExpectJs[/* expect */0](instances));
              }));
        Jest.test("instances - records created instances", (function () {
                var mockFn = jest.fn((function (n) { this.n = n; }));
                Jest.MockJs[/* new1 */1](4, mockFn);
                Jest.MockJs[/* new1 */1](7, mockFn);
                var instances = Jest.MockJs[/* instances */4](mockFn);
                return Jest.ExpectJs[/* toEqual */12](/* array */[
                            {
                              n: 4
                            },
                            {
                              n: 7
                            }
                          ], Jest.ExpectJs[/* expect */0](instances));
              }));
        Jest.test("mockClear - resets calls", (function () {
                var mockFn = jest.fn();
                var before = Jest.MockJs[/* calls */3](mockFn);
                /* tuple */[
                  mockFn.call(/* () */0, 1),
                  mockFn.call(/* () */0, 2)
                ];
                var inbetween = Jest.MockJs[/* calls */3](mockFn);
                mockFn.mockClear();
                var after = Jest.MockJs[/* calls */3](mockFn);
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            /* int array */[],
                            /* int array */[
                              1,
                              2
                            ],
                            /* int array */[]
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                inbetween,
                                after
                              ]));
              }));
        Jest.test("mockReset - resets calls", (function () {
                var mockFn = jest.fn();
                var before = Jest.MockJs[/* calls */3](mockFn);
                /* tuple */[
                  mockFn.call(/* () */0, 1),
                  mockFn.call(/* () */0, 2)
                ];
                var inbetween = Jest.MockJs[/* calls */3](mockFn);
                mockFn.mockReset();
                var after = Jest.MockJs[/* calls */3](mockFn);
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            /* int array */[],
                            /* int array */[
                              1,
                              2
                            ],
                            /* int array */[]
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                inbetween,
                                after
                              ]));
              }));
        Jest.test("mockReset - resets implementations", (function () {
                var mockFn = jest.fn();
                mockFn.mockReturnValue(128);
                var before = mockFn.call(/* () */0, /* () */0);
                mockFn.mockReset();
                var after = mockFn.call(/* () */0, /* () */0);
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            128,
                            undefined
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                after
                              ]));
              }));
        Jest.test("mockImplementation - sets implementation to use for subsequent invocations", (function () {
                var mockFn = jest.fn();
                var before = mockFn.call(/* () */0, 10);
                mockFn.mockImplementation(Pervasives.string_of_int);
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            undefined,
                            "18",
                            "24"
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                mockFn.call(/* () */0, 18),
                                mockFn.call(/* () */0, 24)
                              ]));
              }));
        Jest.test("mockImplementationOnce - queues implementation for one subsequent invocation", (function () {
                var mockFn = jest.fn();
                var before = mockFn.call(/* () */0, 10);
                mockFn.mockImplementationOnce(Pervasives.string_of_int);
                mockFn.mockImplementationOnce((function (a) {
                        return Pervasives.string_of_int((a << 1));
                      }));
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            undefined,
                            "18",
                            "48",
                            undefined
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                mockFn.call(/* () */0, 18),
                                mockFn.call(/* () */0, 24),
                                mockFn.call(/* () */0, 12)
                              ]));
              }));
        Jest.test("mockReturnThis - returns `this` on subsequent invocations", (function () {
                var mockFn = jest.fn();
                var $$this = "this";
                var fn = mockFn.bind($$this);
                var before = fn.call(/* () */0, /* () */0);
                mockFn.mockReturnThis();
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            undefined,
                            $$this,
                            $$this
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                fn.call(/* () */0, /* () */0),
                                fn.call(/* () */0, /* () */0)
                              ]));
              }));
        Jest.test("mockReturnValue - returns given value on subsequent invocations", (function () {
                var mockFn = jest.fn();
                var before = mockFn.call(/* () */0, 10);
                mockFn.mockReturnValue(146);
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            undefined,
                            146,
                            146
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                mockFn.call(/* () */0, 18),
                                mockFn.call(/* () */0, 24)
                              ]));
              }));
        Jest.test("mockReturnValueOnce - queues implementation for one subsequent invocation", (function () {
                var mockFn = jest.fn();
                var before = mockFn.call(/* () */0, 10);
                mockFn.mockReturnValueOnce(29);
                mockFn.mockReturnValueOnce(41);
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            undefined,
                            29,
                            41,
                            undefined
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                mockFn.call(/* () */0, 18),
                                mockFn.call(/* () */0, 24),
                                mockFn.call(/* () */0, 12)
                              ]));
              }));
        return Jest.test("mockClear - resets instances", (function () {
                      var mockFn = jest.fn((function (n) { this.n = n; }));
                      var before = Jest.MockJs[/* instances */4](mockFn);
                      Jest.MockJs[/* new1 */1](4, mockFn);
                      Jest.MockJs[/* new1 */1](7, mockFn);
                      var inbetween = Jest.MockJs[/* instances */4](mockFn);
                      mockFn.mockClear();
                      var after = Jest.MockJs[/* instances */4](mockFn);
                      return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                                  /* array */[],
                                  /* array */[
                                    {
                                      n: 4
                                    },
                                    {
                                      n: 7
                                    }
                                  ],
                                  /* array */[]
                                ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                      before,
                                      inbetween,
                                      after
                                    ]));
                    }));
      }));

describe("fn", (function () {
        Jest.test("calls implementation", (function () {
                var mockFn = jest.fn(Pervasives.string_of_int);
                return Jest.ExpectJs[/* toBe */2]("18", Jest.ExpectJs[/* expect */0](Curry._1(mockFn, 18)));
              }));
        Jest.test("calls - records call arguments", (function () {
                var mockFn = jest.fn(Pervasives.string_of_int);
                Curry._1(mockFn, 74);
                Curry._1(mockFn, 89435);
                var calls = Jest.MockJs[/* calls */3](mockFn);
                return Jest.ExpectJs[/* toEqual */12](/* int array */[
                            74,
                            89435
                          ], Jest.ExpectJs[/* expect */0](calls));
              }));
        it.skip("mockClear - resets calls", (function () {
                var mockFn = jest.fn(Pervasives.string_of_int);
                var before = Jest.MockJs[/* calls */3](mockFn);
                /* tuple */[
                  Curry._1(mockFn, 1),
                  Curry._1(mockFn, 2)
                ];
                var inbetween = Jest.MockJs[/* calls */3](mockFn);
                mockFn.mockClear();
                var after = Jest.MockJs[/* calls */3](mockFn);
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            /* int array */[],
                            /* int array */[
                              1,
                              2
                            ],
                            /* int array */[]
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                inbetween,
                                after
                              ]));
              }));
        Jest.test("mockReset - resets calls", (function () {
                var mockFn = jest.fn(Pervasives.string_of_int);
                var before = Jest.MockJs[/* calls */3](mockFn);
                /* tuple */[
                  Curry._1(mockFn, 1),
                  Curry._1(mockFn, 2)
                ];
                var inbetween = Jest.MockJs[/* calls */3](mockFn);
                mockFn.mockReset();
                var after = Jest.MockJs[/* calls */3](mockFn);
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            /* int array */[],
                            /* int array */[
                              1,
                              2
                            ],
                            /* int array */[]
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                inbetween,
                                after
                              ]));
              }));
        it.skip("mockReset - resets implementations", (function () {
                var mockFn = jest.fn(Pervasives.string_of_int);
                mockFn.mockReturnValue("128");
                var before = Curry._1(mockFn, 0);
                mockFn.mockReset();
                var after = Curry._1(mockFn, 1);
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            "128",
                            "1"
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                after
                              ]));
              }));
        Jest.test("mockImplementation - sets implementation to use for subsequent invocations", (function () {
                var mockFn = jest.fn(Pervasives.string_of_int);
                var before = Curry._1(mockFn, 10);
                mockFn.mockImplementation((function (a) {
                        return Pervasives.string_of_int((a << 1));
                      }));
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            "10",
                            "36",
                            "48"
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                Curry._1(mockFn, 18),
                                Curry._1(mockFn, 24)
                              ]));
              }));
        Jest.test("mockImplementationOnce - queues implementation for one subsequent invocation", (function () {
                var mockFn = jest.fn(Pervasives.string_of_int);
                var before = Curry._1(mockFn, 10);
                mockFn.mockImplementationOnce((function (a) {
                        return Pervasives.string_of_int(Caml_int32.imul(a, 3));
                      }));
                mockFn.mockImplementationOnce((function (a) {
                        return Pervasives.string_of_int((a << 1));
                      }));
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            "10",
                            "54",
                            "48",
                            "12"
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                Curry._1(mockFn, 18),
                                Curry._1(mockFn, 24),
                                Curry._1(mockFn, 12)
                              ]));
              }));
        Jest.test("mockReturnValue - returns given value on subsequent invocations", (function () {
                var mockFn = jest.fn(Pervasives.string_of_int);
                var before = Curry._1(mockFn, 10);
                mockFn.mockReturnValue("146");
                return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                            "10",
                            "146",
                            "146"
                          ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                before,
                                Curry._1(mockFn, 18),
                                Curry._1(mockFn, 24)
                              ]));
              }));
        return Jest.test("mockReturnValueOnce - queues implementation for one subsequent invocation", (function () {
                      var mockFn = jest.fn(Pervasives.string_of_int);
                      var before = Curry._1(mockFn, 10);
                      mockFn.mockReturnValueOnce("29");
                      mockFn.mockReturnValueOnce("41");
                      return Jest.ExpectJs[/* toEqual */12](/* tuple */[
                                  "10",
                                  "29",
                                  "41",
                                  "12"
                                ], Jest.ExpectJs[/* expect */0](/* tuple */[
                                      before,
                                      Curry._1(mockFn, 18),
                                      Curry._1(mockFn, 24),
                                      Curry._1(mockFn, 12)
                                    ]));
                    }));
      }));

describe("fn2", (function () {
        return Jest.test("calls implementation", (function () {
                      var mockFn = jest.fn((function (a, b) {
                              return Pervasives.string_of_int(a + b | 0);
                            }));
                      return Jest.ExpectJs[/* toBe */2]("42", Jest.ExpectJs[/* expect */0](mockFn.call(0, 18, 24)));
                    }));
      }));

describe("MockJs.new", (function () {
        Jest.test("MockJs.new0", (function () {
                var mockFn = jest.fn((function () { this.n = 42; }));
                var instance = Jest.MockJs[/* new0 */0](mockFn);
                return Jest.ExpectJs[/* toEqual */12]({
                            n: 42
                          }, Jest.ExpectJs[/* expect */0](instance));
              }));
        Jest.test("MockJs.new1", (function () {
                var mockFn = jest.fn((function (n) { this.n = n; }));
                var instance = Jest.MockJs[/* new1 */1](4, mockFn);
                return Jest.ExpectJs[/* toEqual */12]({
                            n: 4
                          }, Jest.ExpectJs[/* expect */0](instance));
              }));
        return Jest.test("MockJs.new2", (function () {
                      var mockFn = jest.fn((function (a, b) { this.n = a * b; }));
                      var instance = Jest.MockJs[/* new2 */2](4, 7, mockFn);
                      return Jest.ExpectJs[/* toEqual */12]({
                                  n: 28
                                }, Jest.ExpectJs[/* expect */0](instance));
                    }));
      }));

exports.call = call;
/*  Not a pure module */
